File: RecompSwift/RecompApp/Features/Workouts/MesocycleBanner.swift
import SwiftUI
import RefactorKit

/// Where the lifter is in the current training block.
///
/// Without this, a multi-week program is an undifferentiated wall of sessions. Naming the
/// phase is what makes a lighter week read as strategy rather than as falling behind.
///
/// Mirrors web `MesocycleBanner.tsx`.
struct MesocycleBanner: View {
    let resolution: Mesocycle.Resolution

    private var state: Mesocycle.State { resolution.state }

    private var tint: Color {
        switch state.phase {
        case .accumulation: return .appAccent
        case .peak: return .appSage
        case .deload: return .appWarm
        }
    }

    private var showWarning: Bool {
        resolution.deload.urgency == .soon && !resolution.deloadForced
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 8) {
                Text(state.phase.label.uppercased())
                    .font(.caption2.weight(.bold))
                    .foregroundStyle(tint)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 3)
                    .background(tint.opacity(0.12), in: Capsule())

                Text("Week \(state.weekInBlock) of \(state.blockLength)")
                    .font(.subheadline.weight(.medium))

                Text("Block \(state.blockNumber)")
                    .font(.caption)
                    .foregroundStyle(.secondary)

                if resolution.deloadForced {
                    Text("Pulled forward")
                        .font(.caption2.weight(.semibold))
                        .foregroundStyle(Color.appWarm)
                        .padding(.horizontal, 6)
                        .padding(.vertical, 2)
                        .background(Color.appWarm.opacity(0.15), in: Capsule())
                }

                Spacer(minLength: 0)
            }

            Text(state.summary)
                .font(.caption)
                .foregroundStyle(.secondary)
                .fixedSize(horizontal: false, vertical: true)

            // Week dots make the block's shape legible at a glance.
            HStack(spacing: 4) {
                ForEach(1...max(1, state.blockLength), id: \.self) { week in
                    Capsule()
                        .fill(dotColor(week: week))
                        .frame(height: 5)
                }
            }
            .accessibilityHidden(true)

            if (resolution.deloadForced || showWarning), !resolution.deload.reasons.isEmpty {
                VStack(alignment: .leading, spacing: 3) {
                    Text(resolution.deloadForced ? "Why this week is a deload" : "A deload is coming")
                        .font(.caption.weight(.semibold))
                    ForEach(resolution.deload.reasons, id: \.self) { reason in
                        Text("• \(reason)")
                            .font(.caption2)
                            .foregroundStyle(.secondary)
                            .fixedSize(horizontal: false, vertical: true)
                    }
                }
                .padding(8)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(Color.secondary.opacity(0.08), in: RoundedRectangle(cornerRadius: 8))
            }
        }
        .padding()
        .background(Color.recompSurface, in: RoundedRectangle(cornerRadius: 12))
    }

    private func dotColor(week: Int) -> Color {
        if week == state.weekInBlock { return .appAccent }
        if week < state.weekInBlock { return Color.appAccent.opacity(0.4) }
        if week == state.blockLength { return Color.appWarm.opacity(0.3) }
        return Color.secondary.opacity(0.15)
    }
}

File: recomp-mobile/app/src/main/java/com/refactor/app/ui/workouts/MesocycleBanner.kt
package com.refactor.app.ui.workouts

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Card
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.refactor.app.api.Mesocycle

/**
 * Where the lifter is in the current training block.
 *
 * Without this, a multi-week program is an undifferentiated wall of sessions. Naming the
 * phase is what makes a lighter week read as strategy rather than as falling behind.
 *
 * Mirrors web `MesocycleBanner.tsx` and iOS `MesocycleBanner.swift`.
 */
@Composable
fun MesocycleBanner(
    resolution: Mesocycle.Resolution,
    modifier: Modifier = Modifier,
) {
    val state = resolution.state
    val tint: Color = when (state.phase) {
        Mesocycle.Phase.ACCUMULATION -> MaterialTheme.colorScheme.primary
        Mesocycle.Phase.PEAK -> MaterialTheme.colorScheme.secondary
        Mesocycle.Phase.DELOAD -> MaterialTheme.colorScheme.tertiary
    }
    val showWarning = resolution.deload.urgency == Mesocycle.DeloadUrgency.SOON && !resolution.deloadForced

    Card(modifier = modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp)) {
        Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Text(
                    state.phase.label.uppercase(),
                    style = MaterialTheme.typography.labelSmall,
                    fontWeight = FontWeight.Bold,
                    color = tint,
                    modifier = Modifier
                        .background(tint.copy(alpha = 0.12f), RoundedCornerShape(50))
                        .padding(horizontal = 8.dp, vertical = 3.dp),
                )
                Text(
                    "Week ${state.weekInBlock} of ${state.blockLength}",
                    style = MaterialTheme.typography.bodyMedium,
                    fontWeight = FontWeight.Medium,
                )
                Text(
                    "Block ${state.blockNumber}",
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                if (resolution.deloadForced) {
                    Text(
                        "Pulled forward",
                        style = MaterialTheme.typography.labelSmall,
                        fontWeight = FontWeight.SemiBold,
                        color = MaterialTheme.colorScheme.tertiary,
                        modifier = Modifier
                            .background(
                                MaterialTheme.colorScheme.tertiary.copy(alpha = 0.15f),
                                RoundedCornerShape(50),
                            )
                            .padding(horizontal = 6.dp, vertical = 2.dp),
                    )
                }
            }

            Text(
                state.summary,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            // Week dots make the block's shape legible at a glance.
            Row(
                Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                for (week in 1..state.blockLength) {
                    val dotColor = when {
                        week == state.weekInBlock -> MaterialTheme.colorScheme.primary
                        week < state.weekInBlock -> MaterialTheme.colorScheme.primary.copy(alpha = 0.4f)
                        week == state.blockLength -> MaterialTheme.colorScheme.tertiary.copy(alpha = 0.3f)
                        else -> MaterialTheme.colorScheme.surfaceVariant
                    }
                    Box(
                        Modifier
                            .weight(1f)
                            .height(5.dp)
                            .clip(RoundedCornerShape(3.dp))
                            .background(dotColor),
                    )
                }
            }

            if ((resolution.deloadForced || showWarning) && resolution.deload.reasons.isNotEmpty()) {
                Surface(
                    shape = RoundedCornerShape(8.dp),
                    color = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f),
                ) {
                    Column(Modifier.padding(8.dp), verticalArrangement = Arrangement.spacedBy(3.dp)) {
                        Text(
                            if (resolution.deloadForced) "Why this week is a deload" else "A deload is coming",
                            style = MaterialTheme.typography.labelMedium,
                            fontWeight = FontWeight.SemiBold,
                        )
                        resolution.deload.reasons.forEach { reason ->
                            Text(
                                "• $reason",
                                style = MaterialTheme.typography.labelSmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                    }
                }
            }
        }
    }
}
